tests/test_network_failure_duplicate_messages.py

```diff
@@ -0,0 +1,247 @@
+            file.write("Hello? Is it me you're looking for?a")
+
+        server_root_directory: Path = Path.cwd() / "tests" / "server"
+        server_file_system = ServerFileSystem(server_root_directory=server_root_directory)
+        mock_server_file_system = Mock(wraps=server_file_system)
+        server = Server(
+            server_ip_address=self.SERVER_IP_ADDRESS,
+            server_port_number=self.SERVER_PORT_NUMBER,
+            file_system=mock_server_file_system,
+            invocation_semantics=InvocationSemantics.AT_LEAST_ONCE,
+        )
+        logger.info("SERVER START")
+        server_thread = threading.Thread(target=server.listen_for_messages)
+        server_thread.start()
+
+        fixed_uuid: UUID = uuid4()
+        logger.info("Sending the first message")
+        response_1 = send_message_and_wait_for_reply(
+            message=AppendFileRequest(request_id=fixed_uuid, file_name="appendme.txt", content=b"a"),
+            recipient_ip_address=self.SERVER_IP_ADDRESS,
+            recipient_port_number=self.SERVER_PORT_NUMBER,
+            max_attempts_to_send_message=3,
+            timeout_in_seconds=5,
+        )
+        time.sleep(1)
+        logger.info("Sending the second message")
+        response_2 = send_message_and_wait_for_reply(
+            message=AppendFileRequest(request_id=fixed_uuid, file_name="appendme.txt", content=b"a"),
+            recipient_ip_address=self.SERVER_IP_ADDRESS,
+            recipient_port_number=self.SERVER_PORT_NUMBER,
+            max_attempts_to_send_message=3,
+            timeout_in_seconds=5,
+        )
+        mock_append_file: Mock = mock_server_file_system.append_file
+        server.stop_listening()
+        server_thread.join()
+        assert 2 == mock_append_file.call_count, "Expected server to invoke append file operation twice"
+        assert response_1, "Expected server to respond to first message"
+        assert response_2, "Expected server to respond to second message"
+        same = (
+            open("tests/server/appendme.txt", "rb").read()
+            == open("tests/server/appendme_appended_once.txt", "rb").read() + b"a"
+        )
+        assert same is True, "Expected original file to be different from the appended file"
+
+    def test_network_failure_experiment_6(self):
+        """
+        Network experiment 6: simulate duplicate of append request message.
+
+        Things to test:
+        - Server should invoke the read_file method only once for the first message.
+        - Server should use the message history to respond to the second message.
+
+        innovcation     : at most once
+        operation type  : non-idempotent
+
+        """
+        with open("tests/server/appendme.txt", "w") as file:
+            file.write("Hello? Is it me you're looking for?")
+
+        with open("tests/server/appendme_appended_once.txt", "w") as file:
+            file.write("Hello? Is it me you're looking for?a")
```

Comment:
I don't think we actually need the `appendme_appended_once.txt` file. During assertions later, we can just check that the content of `appendme.txt` is equal to the string `"Hello? Is it me you're looking for?a"`